File: wccloud-common/src/main/java/top/wccloud/common/Result.java
package top.wccloud.common;

import lombok.Data;

import java.io.Serializable;
/**
 * 全局mvc返回bean
 * @author wcz
 */
@Data
public class Result<T> implements Serializable {
    private int code;
    private String msg;

    private T data;

    public Result(int code, String msg, T data) {
        this.code = code;
        this.msg = msg;
        this.data = data;
    }

    public static Result<?> success() {
        return new Result<>(0, null, null);
    }

    public static Result<?> success(String msg) {
        return new Result<>(0, msg, null);
    }

    public static <T> Result<T> success(String msg, T data) {
        return new Result<>(0, msg, data);
    }

    public static Result<?> error(String msg) {
        return new Result<>(-1, msg, null);
    }

    public static <T> Result<T> error(String msg, T data) {
        return new Result<>(-1, msg, data);
    }
}

File: wccloud-web-rust/src/infrastructure/util/result.rs
//! author wcz
use actix_web::body::BoxBody;
use actix_web::http::header::ContentType;
use actix_web::{HttpRequest, HttpResponse, Responder};
use redis::{FromRedisValue, ParsingError, Value};
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
pub struct ResultVO<T: Serialize> {
    pub code: i8,
    pub msg: String,
    pub data: T,
}

#[allow(dead_code)]
pub trait SuccessMsgData<T: Serialize> {
    fn success(msg: String, data: T) -> ResultVO<T> {
        ResultVO { code: 0, msg, data }
    }
}

pub trait SuccessData<T: Serialize> {
    fn success(data: T) -> ResultVO<T> {
        ResultVO {
            code: 0,
            msg: "".to_string(),
            data,
        }
    }
}

impl<T: Serialize> SuccessMsgData<T> for ResultVO<T> {}

impl<T: Serialize> SuccessData<T> for ResultVO<T> {}

impl<T: Serialize> Responder for ResultVO<T> {
    type Body = BoxBody;
    fn respond_to(self, _req: &HttpRequest) -> HttpResponse<Self::Body> {
        HttpResponse::Ok()
            .content_type(ContentType::json())
            .body(serde_json::to_string_pretty(&self).unwrap())
    }
}

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct PageResultVO<T> {
    pub list: Vec<T>,
    pub total: u64,
}

impl<T: for<'a> Deserialize<'a>> FromRedisValue for PageResultVO<T> {
    fn from_redis_value(value: Value) -> Result<Self, ParsingError> {
        let bytes = match value {
            Value::BulkString(vec_bytes) => vec_bytes,
            Value::Nil => return Err("Received nil".into()),
            _ => return Err("Invalid value type".into()),
        };
        let s = String::from_utf8(bytes).map_err(|e| ParsingError::from(e))?;
        serde_json::from_str(&s).map_err(|e| ParsingError::from(e.to_string()))
    }
}

impl<T> PageResultVO<T> {
    pub fn new(list: Vec<T>, total: u64) -> PageResultVO<T> {
        PageResultVO { list, total }
    }
}
